### PR TITLE
[BUGFIX] Fix incorrect type hint and correct typographical errors in DomainBuilder docstrings and fill in missing docstrings

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope
* Fill in missing docstrings and correct docstrings containing typographical errors (due to copy/paste) in `DomainBuilder` classes.
* Also fix incorrect type hint.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-761/GREAT-1285/DX-392
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py
Comment (Kilo59) [inline great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:1]: This file isn't being type-checked
https://github.com/great-expectations/great_expectations/blob/df0ba079b9d3ed533d3df514c01eb30e17cebeef/pyproject.toml#L181

Can you fix the remaining type issues on this module before merging?

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:1]: @Kilo59 My fix was in reaction to PyCharm flagging the error.  I will see if I can make this module subject to type checking.  Thanks.

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:1]: @Kilo59 Please take a look.  Thanks.

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py [hunk 1]
```diff
@@ -35,10 +47,12 @@ class CategoricalColumnDomainBuilder(ColumnDomainBuilder):
     This DomainBuilder uses column cardinality to identify domains.
```
Comment (Kilo59) [inline great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:47]: Can you move these to the if TYPE_CHECKING block?
```python
from great_expectations.validator.computed_metric import MetricValue
from great_expectations.rule_based_profiler.parameter_container import ParameterContainer
```

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py
Comment (Kilo59) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:1]: Thank you!

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py [hunk 2]
```diff
@@ -305,36 +324,36 @@ def _get_domains(
...
+            table_column_name_to_inferred_semantic_domain_type_map=self.semantic_type_filter.table_column_name_to_inferred_semantic_domain_type_map,  # type: ignore[operator,union-attr] # could be None
         )
 
         return domains
 
     def _generate_metric_configurations_to_check_cardinality(
         self,
-        batch_ids: List[str],
         column_names: List[str],
+        batch_ids: Optional[List[str]] = None,
```
Comment (Kilo59) [inline great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:339]: This seems like a somewhat dangerous move. Changing the order of the parameters and changing one from a required positional arg to an optional keyword argument.

If rule_based_profiler was more heavily type-checked I would feel better about this, but right now a lot of the package is being excluded.

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:339]: @Kilo59 This is a private method -- and while most of the package is excluded, this module has been moved along quite a bit.  We should accept inching toward improvements, right?

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:47]: @Kilo59 Done -- thanks.

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py (hunk 2 above)
Comment (Kilo59) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py:339]: I don't know this part of the codebase.
If this is only used within the module or you are reasonably sure of the safety of this move then I'm fine with it.

**Review by Kilo59 — APPROVED** [review]
Pending type-checking issues are resolved.

File: great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py [hunk 3]
```diff
@@ -17,16 +29,16 @@
...
 from great_expectations.validator.metric_configuration import MetricConfiguration
 
 if TYPE_CHECKING:
     from great_expectations.data_context.data_context.abstract_data_context import (
         AbstractDataContext,
     )
+    from great_expectations.rule_based_profiler.parameter_container import (
+        ParameterContainer,  # noqa: TCH001
+    )
+    from great_expectations.validator.computed_metric import MetricValue  # noqa: TCH001
```
Comment (Kilo59) [inline great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py]:
```suggestion
        ParameterContainer
    )
    from great_expectations.validator.computed_metric import MetricValue
```

Please remove these comments

Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/categorical_column_domain_builder.py]: @Kilo59 Committed -- thanks.

**Review by anthonyburdi — APPROVED** [review]

**Pull request merged**